src/HttpCache/VarnishXKeyPurger.php

```diff
@@ -0,0 +1,104 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\HttpCache;
+
+use GuzzleHttp\ClientInterface;
+
+/**
+ * Purges Varnish XKey.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ *
+ * @experimental
+ */
+final class VarnishXKeyPurger implements PurgerInterface
+{
+    private const VARNISH_MAX_HEADER_LENGTH = 8000;
+
+    private $clients;
+    private $maxHeaderLength;
+    private $xkeyGlue;
+
+    /**
+     * @param ClientInterface[] $clients
+     */
+    public function __construct(array $clients, int $maxHeaderLength = self::VARNISH_MAX_HEADER_LENGTH, string $xkeyGlue = ' ')
+    {
+        $this->clients = $clients;
+        $this->maxHeaderLength = $maxHeaderLength;
+        $this->xkeyGlue = $xkeyGlue;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function purge(array $iris)
+    {
+        if (!$iris) {
+            return;
+        }
+
+        $irisChunks = array_chunk($iris, \count($iris));
+
+        foreach ($irisChunks as $irisChunk) {
+            $this->purgeRequest($irisChunk);
+        }
+    }
+
+    private function purgeRequest(array $iris)
+    {
+        foreach ($this->chunkKeys($iris) as $keys) {
+            $this->purgeKeys($keys);
+        }
+    }
+
+    private function purgeKeys(string $keys): void
+    {
+        foreach ($this->clients as $client) {
+            $client->request('PURGE', '', ['headers' => ['xkey' => $keys]]);
+        }
+    }
+
+    private function chunkKeys(array $keys): iterable
+    {
+        $concatenatedKeys = implode($this->xkeyGlue, $keys);
+
+        // If all keys fit in the header, we can return them
+        if (\strlen($concatenatedKeys) <= $this->maxHeaderLength) {
+            yield $concatenatedKeys;
+
+            return;
+        }
+
+        $currentHeader = '';
+
+        foreach ($keys as $position => $key) {
+            // Key can be added to header
+            if (\strlen(sprintf('%s%s%s', $currentHeader, $position > 0 ? $this->xkeyGlue : '', $key)) <= $this->maxHeaderLength) {
+                $currentHeader .= sprintf('%s%s', $position > 0 ? $this->xkeyGlue : '', $key);
+            } else {
+                if (\strlen($key) > $this->maxHeaderLength) {
+                    throw new \Exception(sprintf('Path: %s is too long to fit current max header length (currently set to %s). You must increase max header length.', $key, $this->maxHeaderLength));
```

Comment:
But how does the user increase it?

Reply:
there is a configuration for this

Reply:
You're right. Maybe it would be nice to tell the user which configuration key it is then?
```suggestion
                    throw new \Exception(sprintf('Path: %s is too long to fit current max header length (currently set to %s). You must increase it in http_cache/invalidation/max_header_length configuration.', $key, $this->maxHeaderLength));
```